aquadoggo/src/db/stores/query.rs

```diff
@@ -0,0 +1,724 @@
+            {group}
+
+            -- We always order the rows by document id and list_index, but there might also be
+            -- user-defined ordering on top of that
+            {order}
+
+            -- Connected to cursor pagination we limit the number of rows
+            {limit}
+        "#
+        );
+
+        println!("{sea_quel}");
+
+        let mut rows: Vec<QueryRow> = query_as::<_, QueryRow>(&sea_quel)
+            .fetch_all(&self.pool)
+            .await
+            .map_err(|err| DocumentStorageError::FatalStorageError(err.to_string()))?;
+
+        // We always query one more row than needed to find out if there's more data. This
+        // information aids the user during pagination
+        let has_next_page = if rows.len() as u64 > page_size {
+            // Remove that last row from final results if it exists
+            rows.pop();
+            true
+        } else {
+            false
+        };
+
+        let start_cursor: DocumentCursor = rows.first().unwrap().into();
+        let end_cursor: DocumentCursor = rows.last().unwrap().into();
+
+        let pagination_data = PaginationData {
+            // @TODO
+            total_count: 0,
+            has_next_page,
+            // @TODO
+            has_previous_page: false,
+            start_cursor,
+            end_cursor,
+        };
+
+        // Finally convert everything into the right format
+        let documents = Self::convert_rows(rows, schema.id());
+
+        Ok((pagination_data, documents))
+    }
+
+    /// Merges all fields from the database into documents.
+    fn convert_rows(
+        rows: Vec<QueryRow>,
+        schema_id: &SchemaId,
+    ) -> Vec<(DocumentCursor, StorageDocument)> {
+        let mut converted: Vec<(DocumentCursor, StorageDocument)> = Vec::new();
+
+        if rows.is_empty() {
+            converted
+        } else {
+            let finalize_document = |row: &QueryRow,
+                                     collected_fields: Vec<DocumentViewFieldRow>|
+             -> (DocumentCursor, StorageDocument) {
+                let fields = parse_document_view_field_rows(collected_fields);
+
+                let document = StorageDocument {
+                    id: row.document_id.parse().unwrap(),
+                    fields: Some(fields),
+                    schema_id: schema_id.clone(),
+                    view_id: row.document_view_id.parse().unwrap(),
+                    author: row.owner.parse().unwrap(),
+                    deleted: row.is_deleted,
+                };
+
+                let cursor: DocumentCursor = row.into();
+
+                (cursor, document)
+            };
+
+            let last_row = rows.last().unwrap().clone();
+
+            let mut current = rows[0].clone();
+            let mut current_fields = Vec::new();
+
+            for row in rows {
+                // We observed a new document coming up in the next row, time to change
+                if current.document_id != row.document_id {
```

Comment:
This needs to be revisited for iterating over documents from (pinned) relation lists, since documents can occur multiple times in a row there. A solution would be to also compare the list index